api/db/migrations/20210409154632_add_column_envoisMultiples_to_campaigns_and_add_willBeImproved_to_campaign_participations_and_update_constraint_campaign_participations_unique_userId_campaignId_willBeImproved.js

```diff
@@ -0,0 +1,39 @@
+const CAMPAIGNS_TABLE = 'campaigns';
+const MULTIPLESENDINGS_COLUMN = 'multipleSendings';
+
+const CAMPAIGNPARTICIPATIONS_TABLE = 'campaign-participations';
+const ISIMPROVED_COLUMN = 'isImproved';
+const USERID_COLUMN = 'userId';
+const CAMPAIGNID_COLUMN = 'campaignId';
+const NEW_CONSTRAINT_NAME = 'campaign_participations_campaignid_userid_isImproved';
+
+exports.up = async function(knex) {
+  await knex.schema.table(CAMPAIGNS_TABLE, function(table) {
+    table.boolean(MULTIPLESENDINGS_COLUMN).defaultTo(false);
+  });
+
+  await knex.schema.table(CAMPAIGNPARTICIPATIONS_TABLE, function(table) {
+    table.boolean(ISIMPROVED_COLUMN).defaultTo(false);
+  });
+
+  await knex.schema.table(CAMPAIGNPARTICIPATIONS_TABLE, function(table) {
+    table.dropUnique([CAMPAIGNID_COLUMN, USERID_COLUMN ]);
+  });
+  return knex.raw(`CREATE UNIQUE INDEX ${NEW_CONSTRAINT_NAME} ON "${CAMPAIGNPARTICIPATIONS_TABLE}" ("${CAMPAIGNID_COLUMN}", "${USERID_COLUMN}" , "${ISIMPROVED_COLUMN}" ) WHERE "${ISIMPROVED_COLUMN}" IS FALSE;`);
```

Comment:
j'ai un doute sur si c'est mieux d'écrire ça ou bien : 
```
return knex.raw(`CREATE UNIQUE INDEX ${NEW_CONSTRAINT_NAME} ON "${CAMPAIGNPARTICIPATIONS_TABLE}" ("${CAMPAIGNID_COLUMN}", "${USERID_COLUMN}" ) WHERE "${ISIMPROVED_COLUMN}" IS FALSE;`);
```

Reply:
@laura-bergoens qu'en penses-tu ?

Reply:
je pense que t'as raison @lisequesnel , j'ai trouvé un exemple qui ressemble un peu à notre cas.

`CREATE UNIQUE INDEX ab_c_null_idx ON my_table (id_A, id_B) WHERE id_C IS NULL;
`

selon le [post](https://dba.stackexchange.com/questions/9759/postgresql-multi-column-unique-constraint-and-null-values
)

Reply:
go for it then

Reply:
**question :** Il y a pas du tout cette règle dans le domaine ? Trop compliqué à vérifier dans un use case ?

Reply:
Là, on ajoute uniquement la contrainte en bdd. Dans le usecase on se posera la question dans l'US dédiée je pense